[MRG] Multi-class Brier Score Loss

aggvarun01 opened the pull request:
#### Reference Issues/PRs
Resolves #16055  


#### What does this implement/fix?
The original formulation for Brier score inherently supports multiclass classification [source](https://en.wikipedia.org/wiki/Brier_score#Original_definition_by_Brier). This is currently absent in scikit-learn, which restricts Brier score to binary classification. This PR implements Brier Score for multi-class classification.

#### Notes/Open Questions
There are two different definitions of Brier Score. The one implemented in scikit-learn, which is only applicable for the binary case is:
<img src="https://render.githubusercontent.com/render/math?math=\frac{1}{N}\sum_{i=1}^{N}(y_{i} - \tilde{y}_{i})^{2}">
(where y_hat is the probability of the positive class)

Whereas, the original, and more universal implementation, that is applicable for both the binary and the multi-class case is:
<img src="https://render.githubusercontent.com/render/math?math=\frac{1}{N}\sum_{i=1}^{N}\sum_{c=1}^{C}(y_{ic} - \tilde{y}_{ic})^{2}">

The range of values for the former is [0,1], whereas for the latter, it is [0,2]. For backwards compatibility, this PR uses the old definition for the binary case, and the new one for the multi-class. However, this implementation seems unintuitive, since the range of values changes. I can see the following workarounds:
1. Leave the PR as is, and add this weird behavior to the docstring (possibly in all caps and bold 🤪)
2. Slowly deprecate the 'old' formula, so that even in the binary case, the range of outputs is [0,2]. This has the risk of breaking backwards compatibility and possibly making a lot of people angry.
3. Take inspiration from [R library](https://mlr3measures.mlr-org.com/reference/mbrier.html) and have a separate multi-class brier score function.

While the current PR implements method 1., I personally lean towards method 3.

Commit:
add multi-class support

Commit:
fix swapped y_true y_prob

Commit:
fix docstring

Commit:
fix docstring

Commit:
fix variable name spelling

Commit:
add tests

Commit:
merge upstream

Commit:
import re

Commit:
fix docstring

Commit:
fix linting

Commit:
fix linting

Commit:
remove unused import

ogrisel commented:
> Take inspiration from R library and have a separate multi-class brier score function.

I think this is the best course of action. Introduce a new `multiclass_brier_score` function with the general formulation and keep the existing implementation untouched for binary classification problems. Cross reference the 2 function for each other docstring. Also when trying to call  `brier_score` on multiclass `y_true`, make sure that the message in the raised `ValueError` directs the user to use `multiclass_brier_score` instead.

If the user decide to use `multiclass_brier_score`  on binary problems, I think this is fine. No need to raise any exception or warning as using the [0, 2] range for binary problems can be considered legits.

Commit:
add multiclass_brier_score_loss

Commit:
add tests

Commit:
fix docstring

Commit:
Merge remote-tracking branch 'upstream/master' into multiclass_brier_score_loss

Commit:
use f-strings

Commit:
fix tests

Commit:
fix error message

Commit:
fix docstring

Commit:
fix linting

aggvarun01 commented:
@ogrisel 

1. Added a new `mutliclass_brier_score`
2. Cross-referenced the two functions in each other's docstrings
3. Changed the error message for `brier_score` to point to `multiclass_brier_score`
4. Added a bunch of tests

Should be ready for a review.

ogrisel commented on sklearn/metrics/tests/test_classification.py:
```diff
@@ -2337,6 +2337,79 @@ def test_brier_score_loss():
+                                    [[0.2, 0.7, 0.1],
+                                     [0.6, 0.2, 0.2],
+                                     [0.6, 0.1, 0.3]]),
+        .41333333)
```
Could you please split this test in 2 tests, one for invalid input and error messages and the other for expected numerical values of valid inputs?

Also could you please add a test that check that perfect predictions lead to 0 Brier and perfectly wrong prediction lead to a Brier of 2. (both for a 2-class and a 3-class classification problem).

aggvarun01 replied:
👍

ogrisel commented on sklearn/metrics/_classification.py:
```diff
@@ -2425,6 +2425,12 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+    case as well. When used for the binary case, `multiclass_brier_score_loss`
+    returns Brier score that is exactly twice of the value returned by this
+    function.
+
```
Please recall the latex equation of the docstring here.

aggvarun01 replied:
I copied the equation for `mutliclass_brier_score_loss` as asked. But the docstring for `brier_score_loss` did not include the alternate (i.e. original) formula, so added that equation to the docstring as well.

ogrisel reviewed (commented):
Thanks for bearing with me, here is another round of review comments:

aggvarun01 committed:
Update sklearn/metrics/_classification.py

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

aggvarun01 committed:
Apply suggestions from code review

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

Commit:
split tests

aggvarun01 commented on sklearn/metrics/_classification.py:
```diff
@@ -2504,3 +2511,133 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+        the probabilities provided are assumed to be that of the
+        positive class. The labels in ``y_pred`` are assumed to be
+        ordered alphabetically, as done by
+        :class:`preprocessing.LabelBinarizer`.
```
As you correctly identified, this implementation of `mutliclass_brier_score_loss` shadows the one from `log_loss`. Seems like that change will have to be made to both the functions.

Both the functions use [LabelBinarizer](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.LabelBinarizer.html) to infer labels, so it seems like any warning/error that we raise should be raised there. Right?

aggvarun01 replied:
On second thought, this is a misuse of `LabelBinarizer`

aggvarun01 commented on sklearn/metrics/_classification.py:
```diff
@@ -2504,3 +2511,133 @@ def brier_score_loss(y_true, y_prob, *, sample_weight=None, pos_label=None):
+        else:
+            raise ValueError(f'The number of classes in labels is different '
+                             f'from that in y_prob. Classes found in '
+                             f'labels: {lb.classes_}')
```
Not sure about the convention here. Should I put this function in `sklearn.metrics._base` or in `sklearn.metrics._classification`?

Commit:
add private function

Commit:
add warning for labels

aggvarun01 commented:
@ogrisel 

Apologize for the late reply.

I implemented the changes you suggested. However, there are still a few things that need a review/approval: :

1. Re: [here](https://github.com/scikit-learn/scikit-learn/pull/18699#discussion_r516270516). I added a warning when the label order is different. I added this to the suggested `_validate_multiclass_probabilistic_prediction` function, so that this warning will be raised for `log_loss` as well.
2. Leaving [this](https://github.com/scikit-learn/scikit-learn/pull/18699#discussion_r516277811) for another PR unless you guys think otherwise.
3. Added `_validate_multiclass_probabilistic_prediction` to `sklearn.metrics._classification` itself. Let me know if that's the best place to put it, or should I move it to `sklearn.metrics._base`.
4. This [test](https://github.com/scikit-learn/scikit-learn/blob/master/sklearn/metrics/tests/test_common.py#L556) fails with the new implementation of log_loss, which only changes where `np.clip` is being used. Turns out this test was supposed to fail, and was passing merely because of a floating point error. The way I verified this was as follows:
```python
from sklearn.metrics import log_loss

y_true = np.array([0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1])
y_pred = np.array([0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 0, 1, 1, 0])

def add_dimension(y_prob):
    y_prob = y_prob[:, np.newaxis]
    y_prob = np.append(1 - y_prob, y_prob, axis=1)
    return y_prob

# This is what the test is doing, which gives different results
print(log_loss(y_pred, y_true))  # 15.542609297195847
print(log_loss(y_true, y_pred))  # 15.542649277067358

# alternate way to run the same test, which gives the same results
print(log_loss(y_true, add_dimension(y_pred)))  # 15.542449377709806
print(log_loss(y_pred, add_dimension(y_true)))  # 15.542449377709806
```
I have removed `log_loss` from `NOT_SYMMETRIC_METRICS` for now. I don't know if symmetry even makes sense for `log_loss`, since it takes probabilities.
